[OM 268] - Exporter has to support sending metrics to a HTTP endpoint - using Otel

mphanias opened the pull request.

mphani committed:
OM-268

initial checkin

mphani committed:
OM-268

separated dynatrace logic to its own file

mphani committed:
OM-268

removed println

mphani committed:
OM-268

modified the metric prefix

mphani committed:
OM-268

refactored added dynatrace labels

mphani committed:
OM-268

mphani committed:
OM-268

added http response process logic

mphani committed:
OM-268

mphani committed:
OM-268

mphani committed:
OM-268

refactored otel-http logic

mphani committed:
OM-268

update executor and metrics - configs

mphani committed:
OM-268

mphani committed:
OM-268

code cleanup

mphani committed:
OM-268

refactored code

mphani committed:
OM-268

updated comments

mphani committed:
OM-268

code cleanup

mphani committed:
OM-268

added comments
updated docker-entrypoint script

mphani committed:
OM-268

added defensive check and removed comments/note

mphani committed:
OM-268

clean-up code

mphani committed:
OM-268

code cleanup

mphani committed:
OM-268 - review feedback

updated otel libraries
fixed to delta-temporality

mphani committed:
OM-268 - review feedback

removed endpoint_type
added grpc and http endpoints

mphani committed:
OM-268

refactored variable names

mphani committed:
OM-268

review feedback

mphani committed:
OM-268

removed tls_enabled this is unnecesary

mphani committed:
OM-268

updated config validation logic

mphani committed:
OM-268

removed unnecessary code

mphani committed:
OM-268

mphani committed:
OM-268

optimization

mphani committed:
OM-268

optimization

mphani committed:
OM-268

fix lint errors

mphani committed:
OM-268

refacted counter logic

mphani committed:
OM-268

removed tls

mphani committed:
OM-268

adjusted temporality

mphani committed:
OM-268

mphani committed:
OM-268

refactored

mphani committed:
OM-268

refactored code

mphani committed:
OM-268

use observable gauge and reverted code of delta logic

mphani committed:
OM-268

mphanias committed:
OM-268

config error if both endpoint and grpc_endpoint are configured

mphanias committed:
OM-268

refactored conditional addition of build, cluster labels

mphanias committed:
OM-268

changed to int64 instead of float

mphanias committed:
OM-268

updated config

mphanias committed:
OM-268

review feedback

mphanias committed:
OM-268

removed unnecessary log

mphanias committed:
OM-268

mphanias committed:
OM-268

corrected string in toml file

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
CONNECTOR-1278

mphanias committed:
OM-268

fixed lint issues

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
Merge branch 'master' into OM-268

sunilvirus committed:
Merge branch 'dev' into OM-268

sunilvirus committed:
Fixed go.sum merge issue.

sunilvirus commented on internal/pkg/config/config.go:
```diff
@@ -200,6 +203,37 @@ func (c *Config) ValidateAndUpdate(md toml.MetaData) {
+		c.validateOtelConfigs()
+	}
+
+	if !md.IsDefined("Agent", "OpenTelemetry", "all_metrics_as_gauges") {
```
Should we not initialize this default before calling Otel config validation ?

Can you group all the Otel related handling in one group. 
If you see the flow in this function, first we start general config, then Otel config, and then tls related keys and then again Otel configs. .

mphanias committed:
OM-268 - review feedback

sunilvirus commented on internal/pkg/statprocessors/sp_host_systeminfo.go:
```diff
@@ -6,12 +6,17 @@ import (
 	log "github.com/sirupsen/logrus"
 
 	"github.com/aerospike/aerospike-prometheus-exporter/internal/pkg/commons"
+	config "github.com/aerospike/aerospike-prometheus-exporter/internal/pkg/config"
```
What is the need of alias (with the same name) ?

mphanias replied:
agree, it does not add any benefit, corrected in 3 files.

mphanias committed:
OM-268 - review feedback

mphanias committed:
Merge branch 'master' into OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

refactored

mphanias committed:
OM-268

fixed nil pointer reference

mphanias committed:
OM-268

refactored

mphanias committed:
OM-268

updated test-cases

mphanias committed:
OM-268

fixed user test-cases

mphanias committed:
OM-268

code cleanup, removed println

mphanias committed:
OM-268

remove unnecessary prints

mphanias committed:
OM-268

added is-healthy behaviour

mphanias committed:
OM-268

mphanias committed:
OM-268

removed unnecessary prints

mphanias committed:
OM-268

removed global meter-provider and meters

mphanias committed:
OM-268

moved common maps to object level and updated test-cases

mphanias committed:
OM-268

fixed nil exception

mphanias committed:
OM-268

refactored and removed common and global variables

mphanias committed:
OM-268

moved last 3 common variables as instance variables and updated test-cases

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268

updated comments

mphanias committed:
OM-268

removed unnecessary file

mphanias committed:
OM-268

refactored code to send only node-up metric and not send any other metric when exporter unable to connect to aerospike.

mphanias committed:
OM-268

removed unnecessary condition

mphanias committed:
OM-268

fix to reset refresh-counter to 0 when server errors, to avoid large record / delta push post re-connection

mphanias committed:
Merge branch 'master' into OM-268

sunilvirus commented on internal/pkg/dataprovider/aero_server.go:
```diff
@@ -6,99 +6,112 @@ import (
 
 // Inherits DataProvider interface
 type AerospikeServer struct {
+	fullHost string
```
why do we need 'fullHost' when you can use 'serverHost' for printing purpose ?

Why do we need the 'isServerConnected' and why is it atomic.Bool ? 
this object is not shared across threads anymore in the new design. 
So, why should this atomic Bool ?

clientPolicy is not unique to each connection. 
So, why bother making copy for each instance ? 
Not a big deal if this is really needed.

sunilvirus commented on internal/pkg/dataprovider/aero_server.go:
```diff
@@ -6,99 +6,112 @@ import (
 
 // Aerospike server interaction related code
 
 const (
```
I did not realize it before. It will be nicer if all the consts are at the top of the file.
Not a correctness issue. but more of a file structure.

sunilvirus commented on internal/pkg/dataprovider/aero_server.go:
```diff
@@ -6,99 +6,112 @@ import (
+	clientPolicy   *aero.ClientPolicy
+	serverHost     *aero.Host
+
+	isServerConnected atomic.Bool
```
This isServerConnected is adding duplication. 
Already aeroConnection.IsConnected() tells if the connection is connected.
Now we have 2 things which determine the state. 
We need to keep these 2 info in sync. else they will give 2 diff answers.

mphanias commented on internal/pkg/dataprovider/aero_server.go:
```diff
@@ -121,68 +134,83 @@ func initAerospikeTLS() *tls.Config {
 
 		// Info request
-		requestInfoResponse, err = asConnection.RequestInfo(infoKeys...)
+		callStartTime = time.Now()
```
yes, I am checking those objects ref and connection time etc., - this should be removed

mphanias committed:
OM-268

mphanias committed:
OM-268 - review feedback

changed some vars to const
updated variable names
standardized provider sharing

mphanias committed:
OM-268

cleaned up data-provider
add object creation functions
removed mock related configs

mphanias committed:
OM-268

added test-main and moved initialize as global call

mphanias committed:
OM-268

mphanias committed:
OM-268

mphanias committed:
OM-268 - code hygeine

fixed lint message to use range
fixed log to send to files instead of standard output

mphanias committed:
Merge branch 'dev' into OM-268

mphanias committed:
OM-268

updated go.mod and go.sum

mphanias committed:
Merge branch 'dev' into OM-268

mphanias committed:
OM-268

merge changes from dev & master

mphanias committed:
Merge branch 'dev' into OM-268

mphanias committed:
OM-268

updated libraries merged from master/dev

mphanias committed:
Merge branch 'dev' into OM-268

mphanias committed:
OM-268

fixed empty label issue
removed un-used function params

mphanias committed:
OM-268

handled empty namespace value for batch-index latency

mphanias committed:
Merge branch 'dev' into OM-268

mphanias committed:
OM-268

resolved conflicts from dev branch

mphanias committed:
OM-268

added comments
removed unused variables

mphanias committed:
Merge branch 'dev' into OM-268

tanmayja commented on internal/pkg/statprocessors/statsrefresh.go:
```diff
@@ -63,17 +114,21 @@ func Refresh() ([]AerospikeStat, error) {
+	sr.sharedState.ClusterName = passTwoResponse[sr.sharedState.Infokey_ClusterName]
+	sr.sharedState.Service = passTwoResponse[sr.sharedState.Infokey_Service]
+	sr.sharedState.Build = passTwoResponse[sr.sharedState.Infokey_Build]
+	sr.sharedState.NodeId = passTwoResponse[sr.sharedState.Infokey_NodeId]
```
will it have some value in passTwoResponse for nodeid if removed from infoKeys above?

tanmayja commented on internal/pkg/config/config.go:
```diff
@@ -187,35 +183,71 @@ func (c *Config) ValidateAndUpdate(md toml.MetaData) {
 
 		for k, v := range cloudLabels {
+
 			if v == "" || len(v) == 0 {
```
`if v == "" ` This much is sufficient right?

mphanias commented on internal/pkg/executors/otel_metrics.go:
```diff
@@ -51,61 +52,212 @@ func getCommonLabels() []attribute.KeyValue {
 }
 
-func makeOtelCounterMetric(meter metric.Meter, ctx context.Context, metricName string, desc string, labels []attribute.KeyValue, value float64) {
+func (oe *OtelExecutor) getGaugeMetric(key string, meter metric.Meter, metricName string,
```
I will check, sunil and i debated about this and to keep it less complex to avoid the conditions, separated those 2 functions. will discuss with you

mphanias committed:
OM-268

fixed bugs after code reivew

mphanias committed:
OM-268

fixed bug, added check to have atleast prom or otel enabled.

mphanias committed:
OM-268

fixed bug, atleast prom or otel should be enabled.

mphanias committed:
OM-268

corrected error message

mphanias committed:
OM-268

fixed node_up metric name.

mphanias committed:
OM-268

fixed code review comments

mphanias committed:
OM-268

updated push interval default value to 15

mphanias committed:
OM-269 - review feedback

fixed bug while sending node_up and
corrected use of fatalf, warnf, debugf, infof where formatting is required only
removed un-used function params

mphanias committed:
OM-268 - review feedback

updated temporality constant to follow common pattern

mphanias committed:
OM-268

made metric prefix and metric context-seperator configurable to make it suitable for vairous cloud providers

mphanias committed:
OM-268

fixed config values where quotes are missing for strings

mphanias committed:
OM-268

added key for node_up
fixed metric-gen for node_up as well to use config

mphanias committed:
OM-268

added config for label renaming instead of fixed in codes

mphanias committed:
OM-268

added map creation if renamed_labels is not defined

mphanias committed:
OM-268

corrected node_up metric to use corrected label configs

mphanias committed:
OM-268

fixed node_up label naming use

mphanias committed:
OM-268

corrected cluster label

mphanias committed:
Merge branch 'dev' into OM-268

mphanias committed:
OM-268

updated otel sdk libraries

mphanias committed:
OM-268

updated template config and docket-entrypoint with new env variables.

mphanias committed:
OM-268

fixed review comments, converted configured values to lower-case
updated docker-entrypoint

mphanias committed:
OM-268

removed many unused constants

mphanias committed:
OM-268

updated README.md

mphanias commented on internal/pkg/commons/constants.go:
removed many un-used constants, left couple of them for future use

mphanias committed:
Merge branch 'dev' into OM-268
